# Before you start:
- Read the README.md file
- Comment as much as you can and use the resources (README.md file)
- Happy learning!

In [2]:
# import numpy and pandas
import pandas as pd 
import numpy as np

# Challenge 1 - Exploring the Data

In this challenge, we will examine all salaries of employees of the City of Chicago. We will start by loading the dataset and examining its contents. Please, load the data using Ironhack's database (db: employees, table: employees_advanced).

In [5]:
employee = pd.read_csv('Current_Employee_Names__Salaries__and_Position_Titles.csv')


Examine the `salaries` dataset using the `head` function below.

In [7]:
employee.head()


Name                 Job Titles        Department  \
0  ADRIANO,  RACQUEL ANNE                  PARAMEDIC              FIRE   
1       AFFANEH,  MAHIR A             POLICE OFFICER            POLICE   
2       AFOENYI,  PHYLLIS  TITLE V PROGRAM TRAINEE I  FAMILY & SUPPORT   
3         AGAR,  BULENT B        DEPUTY COMMISSIONER       WATER MGMNT   
4         AGEE,  LAWRENCE             POLICE OFFICER            POLICE   

  Full or Part-Time Salary or Hourly  Typical Hours  Annual Salary  \
0                 F           Salary            NaN        72510.0   
1                 F           Salary            NaN        93354.0   
2                 P           Hourly           20.0            NaN   
3                 F           Salary            NaN       132972.0   
4                 F           Salary            NaN        90024.0   

   Hourly Rate  
0          NaN  
1          NaN  
2         13.0  
3          NaN  
4          NaN

We see from looking at the `head` function that there is quite a bit of missing data. Let's examine how much missing data is in each column. Produce this output in the cell below

In [17]:
employee.isnull().sum()

Name                     0
Job Titles               0
Department               0
Full or Part-Time        0
Salary or Hourly         0
Typical Hours        25528
Annual Salary         8174
Hourly Rate          25528
dtype: int64

Let's also look at the count of hourly vs. salaried employees. Write the code in the cell below

In [20]:
employee['Annual Salary'].isnull().sum()


8174

In [21]:
employee['Hourly Rate'].isnull().sum()

25528

What this information indicates is that the table contains information about two types of employees - salaried and hourly. Some columns apply only to one type of employee while other columns only apply to another kind. This is why there are so many missing values. Therefore, we will not do anything to handle the missing values.

There are different departments in the city. List all departments and the count of employees in each department.

In [25]:
employee['Department'].value_counts()


POLICE                 13848
FIRE                    4627
STREETS & SAN           2175
WATER MGMNT             1914
AVIATION                1856
OEMC                    1737
TRANSPORTN              1113
PUBLIC LIBRARY          1088
DAIS                    1033
FAMILY & SUPPORT         656
FINANCE                  544
HEALTH                   467
LAW                      398
CITY COUNCIL             362
BUILDINGS                258
PUBLIC SAFETY ADMIN      244
BUSINESS AFFAIRS         167
HOUSING & ECON DEV       154
COPA                     123
BOARD OF ELECTION        111
MAYOR'S OFFICE           104
INSPECTOR GEN             95
CITY CLERK                88
PROCUREMENT               84
HUMAN RESOURCES           75
ANIMAL CONTRL             75
CULTURAL AFFAIRS          74
HOUSING                   66
BUDGET & MGMT             44
ADMIN HEARNG              38
DISABILITIES              28
TREASURER                 28
HUMAN RELATIONS           17
BOARD OF ETHICS            8
POLICE BOARD  

# Challenge 2 - Hypothesis Tests

In this section of the lab, we will test whether the hourly wage of all hourly workers is significantly different from $30/hr. Import the correct one sample test function from scipy and perform the hypothesis test for a 95% two sided confidence interval.

In [ ]:
# import scipy.stats
# import numpy as np
# from scipy.stats import ttest_1samp

# confidence_level = 0.95
# degrees_freedom = len(employee) - 1  # or, for large values of sample size, just use that
# sample_mean = np.mean(employee)
# # note that we use the standard error of the sample 
# # as an estimate of the standard error of the population (which is used in the theoretical formula)
# sample_standard_error = scipy.stats.sem(employee) # sem = standard error of the mean = std(mean)/sqrt(samplesize)

# confidence_interval = scipy.stats.t.interval(confidence_level, 
#                                              degrees_freedom, 
#                                              sample_mean, 
#                                              sample_standard_error)

In [47]:
import numpy as np
from scipy.stats import ttest_1samp
employee['Hourly Rate'].mean()

#H0: Hourly Salary = 30
#H1: Hourly Salary != 30 



# TWO SIDED TEST 
sel_hour = employee[~employee['Hourly Rate']. isna()]

#sel_hour['Hourly Rate']


stat, pval = ttest_1samp(sel_hour['Hourly Rate'], 30)
stat
pval

print('stat is  ', stat)
print('pvalue for the one-tailed test is ', pval)



stat is   31.211500329216044
pvalue for the one-tailed test is  3.825475847673794e-202


p-value < 0.05 (smaller than sognificance level (0.05))

we can reject H0:  that means the hourly wage of all hourly workers is significantly different from $30/hr

We are also curious about salaries in the police force. The chief of police in Chicago claimed in a press briefing that salaries this year are higher than last year's mean of $86000/year a year for all salaried employees. Test this one sided hypothesis using a 95% confidence interval.

Hint: A one tailed test has a p-value that is half of the two tailed p-value. If our hypothesis is greater than, then to reject, the test statistic must also be positive.

In [55]:

#Extract relevant data: 
pol = employee[employee['Department']=='POLICE']
#pol
pol_salary = pol['Annual Salary']
#pol_salary


pol['Annual Salary'].isnull().sum()


24

There are NaN values 

In [64]:
sel_pol= pol[~pol['Annual Salary']. isna()]
sel_pol['Annual Salary'].isna().sum()

0

In [83]:
# H0: alary > 86000
# H1: Salary not higher than 86000S

stat, pval = ttest_1samp(sel_pol['Annual Salary'], 86000)
print('stat is  ', stat)
print('pvalue for the one-tailed test is ', pval/2)



stat is   18.51748561981861
pvalue for the one-tailed test is  6.119566952121482e-76


p-value < 0.05 so we can reject H0, that means the salary is not higher, than last year 

Using the `crosstab` function, find the department that has the most hourly workers. 

In [75]:
data_crosstab = pd.crosstab(employee['Department'], (~employee['Hourly Rate'].isna()), margins = False) #margins = True gives totals columns
data_crosstab

Hourly Rate          False  True
Department                      
ADMIN HEARNG            38     0
ANIMAL CONTRL           63    12
AVIATION               642  1214
BOARD OF ELECTION      111     0
BOARD OF ETHICS          8     0
BUDGET & MGMT           42     2
BUILDINGS              258     0
BUSINESS AFFAIRS       160     7
CITY CLERK              88     0
CITY COUNCIL           288    74
COPA                   123     0
CULTURAL AFFAIRS        64    10
DAIS                   288   745
DISABILITIES            27     1
FAMILY & SUPPORT       299   357
FINANCE                499    45
FIRE                  4627     0
HEALTH                 465     2
HOUSING                 66     0
HOUSING & ECON DEV     151     3
HUMAN RELATIONS         17     0
HUMAN RESOURCES         75     0
INSPECTOR GEN           95     0
LAW                    350    48
LICENSE APPL COMM        1     0
MAYOR'S OFFICE          87    17
OEMC                   660  1077
POLICE               13824    24
POLICE BOARD             2     0
PROCUREMENT             82     2
PUBLIC LIBRARY         732   356
PUBLIC SAFETY ADMIN    196    48
STREETS & SAN          329  1846
TRANSPORTN             387   726
TREASURER               28     0
WATER MGMNT            356  1558

STREETS & SAN is the one that has more hourly employees 1846

The workers from the department with the most hourly workers have complained that their hourly wage is less than $35/hour. Using a one sample t-test, test this one-sided hypothesis at the 95% confidence level.

In [79]:
#Extract relevant data: 
s_s = employee[employee['Department']=='STREETS & SAN']
#s_s
s_s_hourly= s_s['Hourly Rate']



s_s_hourly.isnull().sum()
s_s_hourly

15       40.78
16       38.52
26       38.52
49       29.96
65       37.45
         ...  
33624    37.45
33667    37.45
33668    38.52
33676    25.95
33682      NaN
Name: Hourly Rate, Length: 2175, dtype: float64

There are NaN values


In [80]:
s_s_hourly= s_s_hourly[~s_s_hourly. isna()]
s_s_hourly.isna().sum()

0

In [85]:
# H0: hourly wage of STREETS & SAN < 35/hour
# H1: hourly wage of STREETS & SAN >= 35/hour

stat, pval = ttest_1samp(s_s_hourly, 35)
print('stat is  ', stat)
print('pvalue for the one-tailed test is ', pval/2)




stat is   4.000138803968834
pvalue for the one-tailed test is  3.2902372218867895e-05


p-value < 0.05 so we can reject H0, that means the hourly rate in STREETS & SAN is not less than 35 an hour

# Challenge 3: To practice - Constructing Confidence Intervals

While testing our hypothesis is a great way to gather empirical evidence for accepting or rejecting the hypothesis, another way to gather evidence is by creating a confidence interval. A confidence interval gives us information about the true mean of the population. So for a 95% confidence interval, we are 95% sure that the mean of the population is within the confidence interval. 
).

To read more about confidence intervals, click [here](https://en.wikipedia.org/wiki/Confidence_interval).


In the cell below, we will construct a 95% confidence interval for the mean hourly wage of all hourly workers. 

The confidence interval is computed in SciPy using the `t.interval` function. You can read more about this function [here](https://docs.scipy.org/doc/scipy-0.14.0/reference/generated/scipy.stats.t.html).

To compute the confidence interval of the hourly wage, use the 0.95 for the confidence level, number of rows - 1 for degrees of freedom, the mean of the sample for the location parameter and the standard error for the scale. The standard error can be computed using [this](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.sem.html) function in SciPy.

In [88]:
# HOURLY WORKERS:
sel_hour = employee[~employee['Hourly Rate']. isna()]
hourly_workers = sel_hour['Hourly Rate']
hourly_workers

2        13.00
5        37.45
15       40.78
16       38.52
21       34.98
         ...  
33676    25.95
33678    43.72
33681    51.43
33684    49.35
33688    37.45
Name: Hourly Rate, Length: 8174, dtype: float64

In [96]:
import scipy.stats

confidence_level = 0.95
degrees_freedom = len(hourly_workers) - 1 
sample_mean = np.mean(hourly_workers)

sample_standard_error = scipy.stats.sem(hourly_workers) 

confidence_interval = scipy.stats.t.interval(confidence_level, 
                                             degrees_freedom, 
                                             sample_mean, 
                                             sample_standard_error)
confidence_interval

(34.29839539204361, 34.874504045197305)

Now construct the 95% confidence interval for all salaried employeed in the police in the cell below.

In [97]:
# POLICE SALARIED, calculated previously 
pol_annual = sel_pol['Annual Salary']

pol_annual

1         93354.0
4         90024.0
6        111444.0
7        103932.0
9         95736.0
           ...   
33696     80016.0
33697     72510.0
33698     76266.0
33699     90024.0
33700     93354.0
Name: Annual Salary, Length: 13824, dtype: float64

In [98]:
confidence_level = 0.95
degrees_freedom = len(pol_annual) - 1  
sample_mean = np.mean(pol_annual)

sample_standard_error = scipy.stats.sem(pol_annual) 

confidence_interval = scipy.stats.t.interval(confidence_level, 
                                             degrees_freedom, 
                                             sample_mean, 
                                             sample_standard_error)
confidence_interval

(88534.1182885883, 89134.11955863392)

# Bonus Challenge - Hypothesis Tests of Proportions

Another type of one sample test is a hypothesis test of proportions. In this test, we examine whether the proportion of a group in our sample is significantly different than a fraction. 

You can read more about one sample proportion tests [here](http://sphweb.bumc.bu.edu/otlt/MPH-Modules/BS/SAS/SAS6-CategoricalData/SAS6-CategoricalData2.html).

In the cell below, use the `proportions_ztest` function from `statsmodels` to perform a hypothesis test that will determine whether the number of hourly workers in the City of Chicago is significantly different from 25% at the 95% confidence level.

In [13]:
# Your code here:

